web/subpath.go

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package web
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/mattermost-server/mlog"
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/utils"
+)
+
+// UpdateAssetsSubpath rewrites assets in the /client directory to assume the application is hosted
+// at the given subpath instead of at the root. No changes are written unless necessary.
+func UpdateAssetsSubpath(subpath string) error {
+	if subpath == "" {
+		subpath = "/"
+	}
+
+	staticDir, found := utils.FindDir(model.CLIENT_DIR)
+	if !found {
+		return errors.New("failed to find client dir")
+	}
+
+	staticDir, err := filepath.EvalSymlinks(staticDir)
+	if err != nil {
+		return errors.Wrapf(err, "failed to resolve symlinks to %s", staticDir)
+	}
+
+	rootHtmlPath := filepath.Join(staticDir, "root.html")
+	oldRootHtml, err := ioutil.ReadFile(rootHtmlPath)
+	if err != nil {
+		return errors.Wrap(err, "failed to open root.html")
+	}
+
+	pathToReplace := "/static/"
+	newPath := path.Join(subpath, "static") + "/"
```

Comment:
Could `path.Join()` give us the wrong delimiter on certain OSes? Generally we only run on some flavour of linux but we do unofficially support windows

Reply:
In this case, I think `path` is the right library to use since I'm effectively operating on URLs inside the files, vs. the actual paths to the files themselves. It's definitely fun juggling both `filepath` and `path` in this context :)